pkg/server/ca/manager/journal.go

```diff
@@ -25,61 +31,69 @@ const (
 	journalPEMType = "SPIRE CA JOURNAL"
 )
 
-type JournalEntries = journal.Entries
-type X509CAEntry = journal.X509CAEntry
-type JWTKeyEntry = journal.JWTKeyEntry
+type journalEntries = journal.Entries
+type x509CAEntry = journal.X509CAEntry
+type jwtKeyEntry = journal.JWTKeyEntry
+
+type journalConfig struct {
+	cat      catalog.Catalog
+	log      logrus.FieldLogger
+	filePath string
+}
 
 // Journal stores X509 CAs and JWT keys on disk as they are rotated by the
 // manager. The data format on disk is a PEM encoded protocol buffer.
 type Journal struct {
-	path string
+	config *journalConfig
 
-	mu      sync.RWMutex
-	entries *JournalEntries
+	mu                    sync.RWMutex
+	activeX509AuthorityID string
+	caJournalID           uint
+	entries               *journalEntries
 }
 
-func LoadJournal(path string) (*Journal, error) {
-	j := &Journal{
-		path:    path,
-		entries: new(JournalEntries),
-	}
-
-	pemBytes, err := os.ReadFile(path)
-	if err != nil {
-		if os.IsNotExist(err) {
+func LoadJournal(ctx context.Context, config *journalConfig) (*Journal, error) {
+	if fflag.IsSet(fflag.FlagCAJournalInDatastore) {
+		// Look for the CA journal of this server in the datastore.
+		j, err := loadJournalFromDS(ctx, config)
+		if err != nil {
+			return nil, fmt.Errorf("failed to load journal from datastore: %w", err)
+		}
+		if j != nil {
+			// A CA journal record corresponding to this server was found in the
+			// datastore.
 			return j, nil
 		}
-		return nil, errs.Wrap(err)
-	}
-	pemBlock, _ := pem.Decode(pemBytes)
-	if pemBlock == nil {
-		return nil, errs.New("invalid PEM block")
-	}
-	if pemBlock.Type != journalPEMType {
-		return nil, errs.New("invalid PEM block type %q", pemBlock.Type)
 	}
 
-	if err := proto.Unmarshal(pemBlock.Bytes, j.entries); err != nil {
-		return nil, errs.New("unable to unmarshal entries: %v", err)
+	// There is no CA journal record corresponding to this server in the
+	// datastore. Try to load the journal from disk.
+
+	// TODO: stop trying to load the journal from disk in v1.10 and delete
+	// the journal file if exists.
```

Comment:
Moving the journal off the disk is to improve both deployability and security.  If we leave the old journal on disk, it might leak the information necessary to perform a rotation to a CA that is very similar to the one in SPIRE.

In this case, after a successful migration of the Journal into the database, removing the Journal from the disk prevents a second attempt at migration, and greatly reduces the chance that an old journal overwrites the current journal (as we expect that the database journal will always be newer than the disk journal after this commit is merged).

Reply:
We discussed this a little bit during the last maintainer's call and I think that the consensus is that the right thing to do would be to remove the journal from disk in v1.10.